concent_api/gatekeeper/views.py

```diff
@@ -1,11 +1,77 @@
 from base64                         import b64decode
 
+from django.conf                    import settings
 from django.http                    import HttpResponse
-from django.views.decorators.http   import require_POST
 from django.views.decorators.csrf   import csrf_exempt
+from django.views.decorators.http   import require_POST
+from django.views.decorators.http   import require_GET
+
+from gatekeeper.utils               import gatekeeper_access_denied_response
+
+from golem_messages.message         import init_messages
+from golem_messages.message         import Message
+from golem_messages.shortcuts       import load
 
 
 @csrf_exempt
 @require_POST
-def file_transfer_auth(request):
-    return HttpResponse("", status=200)
+def upload(request):
+    if request.content_type != 'application/x-www-form-urlencoded':
+        return gatekeeper_access_denied_response('Unsupported content type.')
+
+    init_messages()
+    response = parse_headers(request)
+    if response is not None:
+        return response
+
+    return HttpResponse("", status = 200)
+
+
+@csrf_exempt
+@require_GET
+def download(request):
+    if request.content_type == 'application/x-www-form-urlencoded':
```

Comment:
Download is not supposed to get any data from the client so why is it checking the `Content-Type` header here? The download request should not have this header at all.

Reply:
When I'm sending curl message with GET it always have:
'CONTENT_TYPE': 'text/plain'
That's why I wanted to have that check.

Reply:
Could you give me a curl command that sends this request? Are you sure it's a download request? The client should not be doing that if there is no body in the request.

For example try this:
``` bash
curl http://google.com -v
```

This is a typical GET request. You can see that there is no `Content-Type` header:
```
> GET / HTTP/1.1
> Host: google.com
> User-Agent: curl/7.57.0
> Accept: */*
```

Reply:
For example:
```
curl -v -H "Authorization: Golem D6UABX47Mo+sAAGmPhp5+S5/iGSvxfrxE+2kCOblP35IcFIGIW4q7es0I1cXWOYs3Tq1SFm5DEPVZj5N5PFWpLqq3d4sgnMRtaGOAQTNrdDEW84chuDsicYhHDfVVGah9Ao33Km1BUhBECZMB38pWTsTnMnu9AvbQ8w+85YX8JnzmUiNuZFkDfnk3ra2klg3WNSZJhpZg+HhmMn5bJqBmtxON+tZ0vJKlws64+W35w7GKYROJ5VLQOp62cQKJamiE+Us8KzulZlLjFobc9o4cKLUQe5FVMuBhNSpk0xVMpoRJgrgXrTPwQjqch+ADyHT88X/0GOkaohfYB2nJCDvrZQmaoCBg1uzQqUktImBwtpiYqzS15HFav8Dk03gCo4dSHTSoqiJtRFCVdFAe1uqUHMcJ9tO8jg2hoXfBVf6YJ5rxULM/52AVnUuDJX4TfrJ8t2Fh0ElHhYY+cTjyC9BRsfWQeLWshyUl3whCRRhM6nxXIHJ69JrwO+gPILri/nI1JArMMRBma9QOtrbF634SYg/2oqn0fRVLTPTi84pQUKiGNlJ8+4gnKD5vPvyWFHPLEd9wiqokkih9veI2MU87Fg8zcwogSSgUd1fhFpDAejqkNtraxVumUIWI2l/pApnkdFPuq7qaMWr8r6dpsb9uYCP7au/UiWzaGFHurGgazBoVTeat78JXkeO52Hhw3QdZJ3z6nxH/HmUhHxW4Nh5iPEPX9Pid8shKaso" -H "Concent-Client-Public-Key: 85cZzVjahnRpUBwm0zlNnqTdYom1LF1P1WNShLg17cmhN2UssnPrCjHKTi5susO3wrr/q07eswumbL82b4HgOw==" -X GET http://127.0.0.1:8000/gatekeeper/download-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend
```
always gives:
```
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /gatekeeper/download-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.47.0
> Accept: */*
> Authorization: Golem D6UABX47Mo+sAAGmPhp5+S5/iGSvxfrxE+2kCOblP35IcFIGIW4q7es0I1cXWOYs3Tq1SFm5DEPVZj5N5PFWpLqq3d4sgnMRtaGOAQTNrdDEW84chuDsicYhHDfVVGah9Ao33Km1BUhBECZMB38pWTsTnMnu9AvbQ8w+85YX8JnzmUiNuZFkDfnk3ra2klg3WNSZJhpZg+HhmMn5bJqBmtxON+tZ0vJKlws64+W35w7GKYROJ5VLQOp62cQKJamiE+Us8KzulZlLjFobc9o4cKLUQe5FVMuBhNSpk0xVMpoRJgrgXrTPwQjqch+ADyHT88X/0GOkaohfYB2nJCDvrZQmaoCBg1uzQqUktImBwtpiYqzS15HFav8Dk03gCo4dSHTSoqiJtRFCVdFAe1uqUHMcJ9tO8jg2hoXfBVf6YJ5rxULM/52AVnUuDJX4TfrJ8t2Fh0ElHhYY+cTjyC9BRsfWQeLWshyUl3whCRRhM6nxXIHJ69JrwO+gPILri/nI1JArMMRBma9QOtrbF634SYg/2oqn0fRVLTPTi84pQUKiGNlJ8+4gnKD5vPvyWFHPLEd9wiqokkih9veI2MU87Fg8zcwogSSgUd1fhFpDAejqkNtraxVumUIWI2l/pApnkdFPuq7qaMWr8r6dpsb9uYCP7au/UiWzaGFHurGgazBoVTeat78JXkeO52Hhw3QdZJ3z6nxH/HmUhHxW4Nh5iPEPX9Pid8shKaso
> Concent-Client-Public-Key: 85cZzVjahnRpUBwm0zlNnqTdYom1LF1P1WNShLg17cmhN2UssnPrCjHKTi5susO3wrr/q07eswumbL82b4HgOw==
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 500 Internal Server Error
< Date: Fri, 05 Jan 2018 14:19:56 GMT
< Server: WSGIServer/0.2 CPython/3.5.2
< Content-Type: text/html
< Content-Length: 121508
< Vary: Cookie
< X-Frame-Options: SAMEORIGIN
```

Reply:
> < HTTP/1.0 500 Internal Server Error
> < Date: Fri, 05 Jan 2018 14:19:56 GMT
> < Server: WSGIServer/0.2 CPython/3.5.2
> < Content-Type: text/html
> < Content-Length: 121508

This is a response header, not a request header.

But after looking into this, looks like you're really getting `text/plain` in `request.content_type` in Django. Even though curl does not send the `Content-Type` header, Django returns `'text.plain'` from `request.content_type`. The WSGI sever must be putting it there as a default. I see that this happens only with `manage.py runserver.` It does not if you run it on `gunicorn` for example. So looks like `runserver` is broken in that regard.

Well, not much we can do about this except hacking around this bug. The whole check is actually more or a sanity check than anything else so it does not have to be perfect. Let's add an exception for `text/plain`. Please make the check accept this one content type and add a comment explaining that this is because of this weird behavior of `runserver`. And that adding this exception does not compromise security because we ignore the body anyway. The check is there just so that the client knows it's sending an incorrect request.

Reply:
Done and pushed on branch gatekeeper-app.